Ports the infinity/interbay vision cone system.

virgili555 opened the pull request:
Mobs are now no longer omniscient and have full 360 degree vision. This only effects other mobs however, items are still fully visible even if they're not in your field of vision for the sake of gameplay simplicity.

AI does not have a vision cone for obvious reasons.

Some pictures of the vision cone in action:
https://imgur.com/a/WnmptGP

virgili555 committed:
Ports the infinity/interbay vision cone system.

virgili555 committed:
Removes unnecessary screen object

virgili555 committed:
Added changelog

PsiOmegaDelta commented on code/_onclick/hud/robot.dm:
```diff
@@ -10,6 +10,18 @@ var/obj/screen/robot_inventory
+		r.vision_cone_overlay.name = ""
+		r.vision_cone_overlay.screen_loc = "1,1"
+		r.vision_cone_overlay.mouse_opacity = 0
+		r.vision_cone_overlay.layer = UNDER_HUD_LAYER
```
Need to also set the plane.

PsiOmegaDelta commented on code/_onclick/hud/human.dm:
```diff
@@ -20,6 +20,17 @@
+		H.vision_cone_overlay.name = ""
+		H.vision_cone_overlay.screen_loc = "1,1"
+		H.vision_cone_overlay.mouse_opacity = 0
+		H.vision_cone_overlay.layer = UNDER_HUD_LAYER
```
Need to also set the plane.

Techhead0 replied:
You should probably just define an `obj/screen` subtype so you don't have to define all these vars every time a vision cone is made.

PsiOmegaDelta commented on code/_onclick/hud/human.dm:
```diff
@@ -20,6 +20,17 @@
+		H.vision_cone_overlay.icon = 'icons/mob/vision_cone.dmi'
+		H.vision_cone_overlay.icon_state = "combat"
+		H.vision_cone_overlay.name = ""
+		H.vision_cone_overlay.screen_loc = "1,1"
```
"SOUTH,WEST"

PsiOmegaDelta commented on code/_onclick/hud/robot.dm:
```diff
@@ -10,6 +10,18 @@ var/obj/screen/robot_inventory
+		r.vision_cone_overlay.icon = 'icons/mob/vision_cone.dmi'
+		r.vision_cone_overlay.icon_state = "combat"
+		r.vision_cone_overlay.name = ""
+		r.vision_cone_overlay.screen_loc = "1,1"
```
"SOUTH,WEST"

PsiOmegaDelta commented on code/modules/mob/living/living.dm:
```diff
@@ -5,6 +5,14 @@
+/mob/living/Initialize()
+	. = ..()
+	if(can_have_vision_cone)
+		vision_cone = 1
```
..why do we have multiple vars for this?

virgili555 replied:
One is "should it have one" and the other is "does have one". 

Lines 51 to 54 in vision_cone.dm use it to remove the vision cone if they have one, and if they shouldn't have one.

PsiOmegaDelta replied:
That seems rather superfluous when you can just check if `vision_cone_overlay` is set.

virgili555 replied:
Possibly. I don't really want to touch it though, because I'd likely break things.

PsiOmegaDelta commented on code/modules/vision_cone/vision_cone.dm:
```diff
@@ -0,0 +1,141 @@
+				M.in_vision_cones[src.client] = 1
+
+	else
+		return
```
These last 2 lines are extremely meaningless.

PsiOmegaDelta commented on code/modules/vision_cone/vision_cone.dm:
```diff
@@ -0,0 +1,141 @@
+
+/mob/living/proc/clear_cone_effect(var/image/I)
+	if(I)
+		qdel(I)
```
qdel() already nullchecks.

PsiOmegaDelta commented on code/modules/vision_cone/vision_cone.dm:
```diff
@@ -0,0 +1,141 @@
+		src.vision_cone = 0
+
+/mob/living/set_dir()
+	..()
```
`. = ..()`

PsiOmegaDelta commented on code/modules/mob/living/living.dm:
```diff
@@ -5,6 +5,14 @@
+	if(can_have_vision_cone)
+		vision_cone = 1
+
+	GLOB.dir_set_event.register(src, src, /mob/proc/update_vision_cone)
```
You're already overriding the `set_dir()`-proc. There's no need to also listen to the event.

PsiOmegaDelta commented on code/modules/vision_cone/vision_cone.dm:
```diff
@@ -0,0 +1,141 @@
+
+/mob/living/set_dir()
+	..()
+	update_vision_cone()
```
For bonus points, only call `update_vision_cone()` if . is actually true.

HeyBanditoz commented on code/modules/mob/living/living.dm:
```diff
@@ -5,6 +5,14 @@
+/mob/living/Initialize()
+	. = ..()
+	if(can_have_vision_cone)
+		vision_cone = 1
```
Think you could do `vision_cone = can_have_vision_cone` instead.

PsiOmegaDelta commented on code/modules/mob/living/living_defines.dm:
```diff
@@ -47,6 +47,11 @@
 	var/stuttering = null	//Carbon
 	var/slurring = null		//Carbon
 
+	var/list/in_vision_cones = list()
```
It's preferable to use LAZYADD, LAZYREMOVE, LAZYLEN, etc. when possible over pre-initializing lists.
There is a global list limit.

HeyBanditoz commented on code/modules/mob/living/silicon/silicon.dm:
```diff
@@ -2,6 +2,7 @@
 	gender = NEUTER
 	voice_name = "synthesized voice"
 	skillset = /datum/skillset/silicon
+	can_have_vision_cone = TRUE
```
I think borgs and such _should_ be able to see 360 degrees around them. It'd be a decent buff to them, at least.

virgili555 replied:
That wouldn't make sense since in their sprites their cameras are always forward facing.

SamSamsonthe3rd replied:
You could easily explain that as just being their visible "face", so they're more appealing to organics. They could have dozens of tiny cameras all over their bodies.

PsiOmegaDelta commented on code/modules/vision_cone/vision_cone.dm:
```diff
@@ -0,0 +1,141 @@
+			for(M in cone(src, OPPOSITE_DIR(src.dir), view(10, src)))
+				I = image("split", M)
+				I.override = 1
+				src.client.images += I
```
These hidden images will likely have to be re-applied when affected client reconnects.
The appearance system can handle this for you.

virgili555 replied:
I've not found any issues with disconnecting/reconnecting relating to the vision cones. Seemed to work as intended during testing.

HeyBanditoz commented on code/modules/vision_cone/vision_cone.dm:
```diff
@@ -0,0 +1,141 @@
+//Defines.
+#define OPPOSITE_DIR(D) turn(D, 180)
+
+/client/
```
Trailing /

mattmatterson111 commented:
Wewlad. 👀

tlc2013 commented:
REJOICE!

...

Whenever that check passes properly.

virgili555 committed:
Put the screen vars into their own object so re-defining each var is no longer necessary.
Addessed a majority of the feedback recieved.
Removed un-used and un-needed code.

mattmatterson111 commented:
I avoided making this PR myself so that someone else could take all the blame for shitty code. Muahahahah

virgili555 committed:
Removed more unnecessary code.
Removed unnecessary event registered/unregisters.

virgili555 commented:
Everything that needs to be addressed should have been addressed now. Hopefully.

comma commented on code/modules/mob/living/living.dm:
```diff
@@ -5,6 +5,9 @@
 		add_to_living_mob_list()
 
+/mob/living/Initialize()
+	. = ..()
```
??? what's this override for

virgili555 committed:
Removes unneeded init override.

PsiOmegaDelta commented on code/modules/vision_cone/vision_cone.dm:
```diff
@@ -49,7 +49,7 @@
 /mob/living/update_vision_cone()
 	if(!can_have_vision_cone)
-		if(vision_cone)
+		if(vision_cone_overlay != null)
```
The `!= null` part is superfluous.

virgili555 committed:
Removes "unnecessary" != check.

virgili555 commented:
Should be good to go now.

babydoll commented:
Needs some images, or ideally videos (webms, gifs)

tlc2013 commented:
Aren't there a couple in the first post? I can't imagine it changing that much with the requested-changes thing.

virgili555 commented:
Yep, images are in the PR. Nothing's changed functionally, just code cleanup.

PsiOmegaDelta commented on code/modules/mob/mob.dm:
```diff
@@ -739,6 +739,8 @@
 	else if( lying != lying_prev )
 		update_icons()
 
+	update_vision_cone()
```
What exactly is done in this proc that affects the vision cone?

virgili555 replied:
Updates the vision cone when canmove and icons are updated. Keeps things pretty much always up to date for the cone (I think?)

mattmatterson111 replied:
It was originally so that whenever you rested it updated the cone to appear and disappear. Because I believe the original code had some check in `update_vision_cone()` to make it so the cone disappears if you lying at all, and it would check that in the canmove thing to appear or disappear depending on if you're lying down or not.

PsiOmegaDelta replied:
Well, currently `update_vision_cone()` has no dependencies on being unable to move, being forced to lie down or otherwise and due to its loops over mobs in range shouldn't be called constantly and needlessly.

PsiOmegaDelta commented on code/modules/mob/living/living.dm:
```diff
@@ -580,6 +580,11 @@ default behaviour is:
+	for(var/mob/M in oview(src))
+		M.update_vision_cone()
+
+	update_vision_cone()
```
The case when mobs are forceMoved is not covered by this.

virgili555 replied:
Non-issue since the vision cones constantly update. You'll see them very briefly but they'll disappear, and that's assuming they don't move at all after being force moved.

PsiOmegaDelta replied:
I'd say it's an issue that the vision cone constantly updates.

Minijar commented:
Will you still hear things that are outside of your cone of vision? If so, won't it be extremely annoying hearing people you can't see, or on the other hand, very bad you can't hear someone stood behind you?

virgili555 commented:
@Minijar You can, yes. And that's kind of the point. Humans shouldn't be omniscient.

mattmatterson111 commented:
Does this implementation include the "noise" icon when someone makes a footstep sound in the cone?

virgili555 commented:
No, that's purposefully not been ported yet as it'd make sneaking unfeasible without other changes, which is slightly outside of what I wanted to achieve with this PR.

PurpleMartinJCK commented on code/modules/mob/living/carbon/human/species/station/nabber.dm:
```diff
@@ -32,6 +32,8 @@
 
 	reagent_tag = IS_NABBER
 
+	have_vision_cone = FALSE
```
Fuck yeah

CrimsonShrike replied:
I pushed for this during infinitystation PR, glad to see it made it here

PurpleMartinJCK commented:
How does this function with dark vision and in darker areas?

virgili555 commented:
The same, just with a vision cone that blocks vision of mobs inside it.

PurpleMartinJCK commented:
So can you see someone behind you approaching with a flashlight?

virgili555 commented:
You'll see the light, but not the person or the stuff they're holding.

Images for reference:
https://imgur.com/a/OeGFzZz

You can still hear their footsteps and if they cough/sneeze/talk/etc

babydoll commented:
At **most** some species should have a wider viewing angle. I'm extremely uninterested in someone having 360 vision while everyone else has much less.

HeyBanditoz commented:
Uh on topic dionea should have 360* I think

virgili555 commented:
I don't know which should and which should not, so I'm leaving it as is for now.

virgili555 commented:
I do not have the time to give this any more development than I already have, especially not without a bit more guidence due to my inexperience. I'll be leaving the branch up on my repo if anyone else wants to/can take over.

comma commented:
Uggh hang on trying to steal it
